README.md
mdown
=====

[![Build Status](https://travis-ci.org/dikmax/mdown.svg?branch=master)](https://travis-ci.org/dikmax/mdown)
[![codecov](https://codecov.io/gh/dikmax/mdown/branch/master/graph/badge.svg)](https://codecov.io/gh/dikmax/mdown)
[![Pub](https://img.shields.io/pub/v/mdown.svg)](https://pub.dartlang.org/packages/mdown)
[![CommonMark spec](https://img.shields.io/badge/commonmark-0.28-green.svg)](http://spec.commonmark.org/)

***mdown*** is fast and [CommonMark][]-compliant Markdown parser.

Basic usage:

```dart
print(markdownToHtml('# Hello world!'));
```

Project main goal is create processing library for Markdown.

Performance
===========

As there are not many Markdown parsers written in Dart out there,
parsing speed is compared with markdown package. [Progit][] was used as
a source of markdown files in different languages. ***mdown*** appears
to be **3** times faster in **VM**, **11** times faster
in **Chrome**, **2.2** times faster in **Safari** and **3.7** times
faster in **Firefox**.

[Run benchmarks yourself or see details.](https://github.com/dikmax/mdown-benchmark)

***mdown*** make extensive use of `String.codeUnitAt` instead
of `RegExp`. So you can see noticeable gain for non-latin languages
(up to &times;38 in Chrome for Japan language).

Extensions
==========

***mdown*** supports some language extensions. You can specify enabled
extensions using options parameter in `markdownToHtml`.

```dart
Options options = const Options(superscript: true);
String res = markdownToHtml('Hello world!\n===', options);
```

There three predefined sets of options:

- `Options.strict`: all extensions, except `rawHtml` are disabled
- `Options.commonmark`: only `smartPunctuation` and `rawHtml` extension
  are enabled.
- `Options.gfm`: `rawHtml`, `tagFilter`, `pipeTables`.
- `Options.defaults`: `smartPunctuation`, `strikeout`, `subscript`,
  `superscript`, `pipeTables`, `texMathDollars`, `rawTex`, `rawHtml`
  are enabled.

To get correspondent parser/writer instance use static getter on class:

```dart
String res = markdownToHtml('Hello world!\n===', Options.strict);
```

If second parameter is not provided, `Options.defaults` is used.

Raw HTML (`Options.rawHtml`)
----------------------------

Allows to include raw HTML blocks. Official CommonMark extension.

Tag filter (`Options.tagFilter`)
--------------------------------

Filters `<textarea>`, `<style>`, `<xmp>`, `<iframe>`, `<noembed>`,
`<noframes>`, `<script>`, `<plaintext>` from HTML output. Works together
with `Options.rawHtml`. Part of [GitHub Flavored Markdown][gfm].

Smart punctuation (`Options.smartPunctuation`)
----------------------------------------------

Smart punctuation is automatic replacement of `...`, `---`, `--`, `"`
and `'` to "…", "—", "–" and curly versions of quote marks accordingly.
It's only official extension to date.

**NOTE:** This extension uses Unicode chars. Make sure that your code
supports it.


Extended attributes for fenced code (`Options.fencedCodeAttributes`)
--------------------------------------------------------------------

Allows fenced code block to have arbitrary extended attributes.

``````md
``` {#someId .class1 .class2 key=value}
code
```
``````

This will be rendered in HTML as

```html
<pre id="someId" class="class1 class2" key="value"><code>code
</code></pre>
```


Extended attributes for headings (`Options.headingAttributes`)
--------------------------------------------------------------

Allows headings to have arbitrary extended attributes.

``````md
# Heading 1 {#someId}

Heading 2 {.someClass}
-------------------
``````

This will be rendered in HTML as

```html
<h1 id="someId">Heading 1</h1>
<h2 class="someClass">Heading 2</h2>
```


Extended attributes for inline code (`Options.inlineCodeAttributes`)
--------------------------------------------------------------------

Adds extended attributes support to inline code.

``````md
`code`{#id .class key='value'}
``````

Extended attributes for links and images (`Options.linkAttributes`)
-------------------------------------------------------------------

Extended attributes for links and images. Both inline and reference
links are supported.

``````md
![](image.jpg){width="800" height="600"}

[test][ref]

[ref]: http://test.com/ {#id}
``````

This will be transformed into:

``````html
<p><img src="image.jpg" alt="" width="800" height="600" /></p>
<p><a href="http://test.com/" id="id">test</a></p>
``````

Strikeout (`Options.strikeout`)
-------------------------------

Strikeouts text (~~like this~~). Just wrap text with double tildes (`~~`).

```md
Strikeouts text (~~like this~~).
```


Subscript (`Options.subscript`)
-------------------------------

Support for subscript (H<sub>2</sub>O). Wrap text with tildes (`~`).

```md
H~2~O
```

Subscript couldn't contain spaces. If you need to insert space into the
subscript, escape space (`\ `).

```md
subscript~with\ spaces~
```


Superscript (`Options.superscript`)
-----------------------------------

Support for superscript (2<sup>2</sup>=4). Wrap text with carets (`^`).

```md
2^2^=4
```

Superscript couldn't contain spaces. If you need to insert space into
superscript, escape space (`\ `).

```md
superscript^with\ spaces^
```


Pipe tables (`Options.pipeTables`)
----------------------------------

Allows to parse tables where cells are separated with vertical bars
(`|`). Compatible with [GitHub table syntax][gfm].

```md
head | cells
-----|------
body | cells
more | cells
```

Also supports cells alignment.

```md
:----|:-----:|----:
left aligned | center aligned | right aligned
```


TeX Math between dollars (`Options.texMathDollars`)
---------------------------------------------------

Anything between two `$` characters will be treated as inline TeX math.
The opening `$` must have a non-space character immediately to its
right, while the closing `$` must have a non-space character immediately
to its left, and must not be followed immediately by a digit. Thus,
`$20,000 and $30,000` won’t parse as math. If for some reason you need
to enclose text in literal `$` characters, backslash-escape them and
they won’t be treated as math delimiters.

Anything between two `$$` will be treated as display TeX math.

HTML writer generates markup for [MathJax][] library. I.e. wraps content
with `\(...\)` or `\[...\]` and additionally wraps it with
`<span class="math inline">` or `<span class="math display">`. If you
need custom classes for `span` you can override them with
`Options.inlineTexMathClasses` and `Options.displayTexMathClasses`.


TeX Math between backslashed `()` or `[]` (`Options.texMathSingleBackslash`)
----------------------------------------------------------------------------

Causes anything between `\(` and `\)` to be interpreted as inline TeX
math and anything between `\[` and `\]` to be interpreted as display
TeX math.

**NOTE 1:** This extension breaks escaping of `(` and `[]`.

**NOTE 2:** This extension is disabled by default.


TeX Math between double backslashed `()` or `[]` (`Options.texMathDoubleBackslash`)
-----------------------------------------------------------------------------------

Causes anything between `\\(` and `\\)` to be interpreted as inline TeX
math and anything between `\\[` and `\\]` to be interpreted as display
TeX math.

**NOTE:** This extension is disabled by default.


Raw TeX (`Options.rawTex`)
--------------------------

Allows to include raw TeX blocks into documents. Right now only
environment blocks are supported. Everything between `\begin{...}` and
`\end{...}` is treated as TeX and passed into resulting HTML as is.


Custom reference resolver
-------------------------

Custom reference resolver may be required when parsing document without
implicitly defined references, for example, Dartdoc.

```dart
/**
 * Throws a [StateError] if ...
 * similar to [anotherMethod], but ...
 */
```

In that case, you could supply parser with the resolver, which should
provide all missing links.

```dart
import 'package:mdown/mdown.dart';
import 'package:mdown/ast/standard_ast_factory.dart';

String library = "mdown";
String version = "0.11.0";
Target linkResolver(String normalizedReference, String reference) {
  if (reference.startsWith("new ")) {
    String className = reference.substring(4);
    return astFactory.target(
        "http://www.dartdocs.org/documentation/$library/$version/index.html#$library/$library.$className@id_$className-",
        null);
  } else {
    return null;
  }
}

String res = markdownToHtml('Hello world!\n===', new Options(linkResolver: linkResolver));
```

[CommonMark]: http://commonmark.org/
[gfm]: https://github.github.com/gfm/
[MathJax]: https://www.mathjax.org/
[Progit]: https://github.com/progit/progit


CHANGELOG.md
# 0.13.0

* Update dependencies for Dart 2.0
* Update CommonMark specification to 0.28

# 0.12.0+2

* Avoid shadowing offset variable (Thanks to @enyo)

# 0.12.0+1

* Bump version to force DartDoc rebuild.

# 0.12.0

* Tables support
* Performance optimization.

# 0.11.0+2

* Update README.

# 0.11.0

* `md_proc` is renamed to `mdown`.
* New AST to speed up processing.

# 0.10.2

* Add new block HTML tags from CommonMark 0.27 spec.

# 0.10.1

* Update rules to latest specification 0.27

# 0.10.0

* New extensions: `fencedCodeAttributes`, `headingAttributes`,
`inlineCodeAttributes` and `linkAttributes`.

# 0.9.0+1

* Update README.

# 0.9.0

* Complete rewrite of parsing engine, which now is
around 2x faster than [markdown](https://pub.dartlang.org/packages/markdown).

# 0.8.2

* Update rules to specification 0.24

# 0.8.1

* Support for raw TeX blocks.

# 0.8.0

* Update rules to latest specification 0.23.
* **BREAKING**: `Header` renamed to `Heading` to conform with spec.


lib/options.dart
library mdown.options;

import 'ast/ast.dart';

/// Link resolver accepts two references and should return [Target] with correspondent link.
/// If link doesn't exists link resolver should return `null`.
///
/// CommonMark defines reference as case insensitive. Use [normalizedReference] when you need reference
/// normalized according to CommonMark rules, or just [reference] if you want to get reference as it
/// written in document.
typedef Target LinkResolver(String normalizedReference, String reference);

/// Default resolver doesn't return any link, so parser parses only explicitly written references.
Target defaultLinkResolver(String normalizedReference, String reference) =>
    null;

/// Parser or writer options. Mostly contains list of enabled extensions.
class Options {
  /// Enables raw html blocks and inlines. This officially supported extension
  /// from CommonMark and thus requires explicit disabling.
  final bool rawHtml;

  /// Enables GFM HTML tags filtering
  final bool tagFilter;

  /// Enables smart punctuation extension. It's automatic replacement of `...`, `---`, `--`, `"` and `'` to `…`, `—`,
  /// `–` and curly versions of quote marks accordingly. It's only official extension to date.
  final bool smartPunctuation;

  /// Allows fenced code block to have arbitrary extended attributes.
  ///
  /// ``````md
  /// ``` {#someId .class1 .class2 key=value}
  /// code
  /// ```
  /// ``````
  ///
  /// This will be rendered in HTML as
  ///
  /// ```html
  /// <pre id="someId" class="class1 class2" key="value"><code>code
  /// </code></pre>
  /// ```
  final bool fencedCodeAttributes;

  /// Allows headings to have arbitrary extended attributes.
  ///
  /// ``````md
  /// # Heading 1 {#someId}
  ///
  /// Heading 2 {.someClass}
  /// -------------------
  /// ``````
  ///
  /// This will be rendered in html as
  ///
  /// ```html
  /// <h1 id="someId">Heading 1</h1>
  /// <h2 class="someClass">Heading 2</h2>
  /// ```
  final bool headingAttributes;

  /// Adds extended attributes support to inline code.
  ///
  /// ``````md
  /// `code`{#id .class key='value'}
  /// ``````
  final bool inlineCodeAttributes;

  /// Extended attributes for links and images. Both inline and reference links are
  /// supported.
  ///
  /// ``````md
  /// ![](image.jpg){width="800" height="600"}
  ///
  /// [test][ref]
  /// ``````
  ///
  /// This will be transformed into:
  ///
  /// ``````html
  /// <p><img src="image.jpg" width="800" height="600"/></p>
  /// <p><a href="http://test.com/" id="id">test</a></p>
  /// ``````
  final bool linkAttributes;

  /// Enables strikeout extension. Strikeout is a text wrapped with double tilde (`~~`). Example:
  ///
  /// ```md
  /// Strikeouts text (~~like this~~).
  /// ```
  final bool strikeout;

  /// Enables subscript extension. Subscript is a text wrapped with tilde (`~`). Example:
  ///
  /// ```md
  /// H~2~O
  /// ```
  ///
  /// Subscript couldn't contain spaces. If you need to insert space into subscript, escape space (`\ `).
  ///
  /// ```md
  /// subscript~with\ spaces~
  /// ```
  final bool subscript;

  /// Enables superscript extension. Superscript is a text wrapped with caret (`^`). Example:
  ///
  /// ```md
  /// 2^2^=4
  /// ```
  ///
  /// Superscript couldn't contain spaces. If you need to insert space into superscript, escape space (`\ `).
  ///
  /// ```md
  /// superscript^with\ spaces^
  /// ```
  final bool superscript;

  /// Enables tex math inside `$` or `$$`.
  ///
  /// Anything between two `$` characters will be treated as TeX math. The
  /// opening `$` must have a non-space character immediately to its right,
  /// while the closing `$` must have a non-space character immediately to its
  /// left, and must not be followed immediately by a digit. Thus,
  /// `$20,000 and $30,000` won’t parse as math. If for some reason you need to
  /// enclose text in literal `$` characters, backslash-escape them and they
  /// won’t be treated as math delimiters.
  final bool texMathDollars;

  /// Causes anything between `\(` and `\)` to be interpreted as inline TeX
  /// math, and anything between `\[` and `\]` to be interpreted as display
  /// TeX math.
  ///
  /// Note: a drawback of this extension is that it precludes escaping `(` and
  /// `[`.
  final bool texMathSingleBackslash;

  /// Causes anything between `\\(` and `\\)` to be interpreted as inline TeX
  /// math, and anything between `\\[` and `\\]` to be interpreted as display
  /// TeX math.
  final bool texMathDoubleBackslash;

  /// Classes added to `span` on rendering inline math into html.
  /// Defaults to `['math', 'inline']`.
  final Iterable<String> inlineTexMathClasses;

  /// Classes added to `span` on rendering display math into html.
  /// Defaults to `['math', 'display']`.
  final Iterable<String> displayTexMathClasses;

  /// Enables raw TeX blocks.
  ///
  /// Everything between \begin{env} \end{env} is treated as TeX. Delimiters
  /// should be placed on separate lines and `env` identifiers should be same.
  final bool rawTex;

  final bool pipeTables;

  /// Custom reference resolver may be required when parsing document without
  /// implicit defined references, for example Dartdoc.
  ///
  /// ```dart
  /// /**
  ///  * Throws a [StateError] if ...
  ///  * similar to [anotherMethod], but ...
  ///  */
  /// ```
  ///
  /// In that case you could supply parser with resolver, which should provide all missing links.
  ///
  /// ```dart
  /// String library = "mdown";
  /// String version = "0.4.0";
  /// Target linkResolver(String normalizedReference, String reference) {
  ///   if (reference.startsWith("new ")) {
  ///     String className = reference.substring(4);
  ///     return new Target("http://www.dartdocs.org/documentation/$library/$version/index.html#$library/$library.$className@id_$className-", null);
  ///   } else {
  ///     return null;
  ///   }
  /// }
  ///
  /// CommonMarkParser parser = new CommonMarkParser(const Options(linkResolver: linkResolver));
  /// Document doc = parser.parse('Hello world!\n===');
  /// String res = HtmlWriter.defaults.write(doc);
  /// ```
  final LinkResolver linkResolver;

  /// Constant constructor with required options.
  const Options(
      {this.rawHtml: true,
      this.tagFilter: false,
      this.smartPunctuation: false,
      this.fencedCodeAttributes: false,
      this.headingAttributes: false,
      this.inlineCodeAttributes: false,
      this.linkAttributes: false,
      this.strikeout: false,
      this.subscript: false,
      this.superscript: false,
      this.texMathDollars: false,
      this.texMathSingleBackslash: false,
      this.texMathDoubleBackslash: false,
      this.rawTex: false,
      this.pipeTables: false,
      this.inlineTexMathClasses: const <String>['math', 'inline'],
      this.displayTexMathClasses: const <String>['math', 'display'],
      this.linkResolver: defaultLinkResolver});

  /// Predefined version of Options. Alongside with [strict] also supports smart
  /// puctuation, which is declared separately
  /// in [CommonMark](http://commonmark.org).
  static const Options commonmark =
      const Options(rawHtml: true, smartPunctuation: true);

  /// Predefined version of Options. Enables lot of useful extensions.
  static const Options defaults = const Options(
      rawHtml: true,
      smartPunctuation: true,
      fencedCodeAttributes: true,
      headingAttributes: true,
      inlineCodeAttributes: true,
      linkAttributes: true,
      strikeout: true,
      subscript: true,
      superscript: true,
      texMathDollars: true,
      rawTex: true,
      pipeTables: true);

  /// GitHub-flavoured markdown defaults
  static const Options gfm =
      const Options(rawHtml: true, tagFilter: true, pipeTables: true);

  /// Predefined strict version of Options. Only support
  /// [CommonMark specification](http://commonmark.org).
  static const Options strict = const Options(rawHtml: true);
}
